### ISSUE TITLE
CFDE: Save and reuse recordset queries

### ISSUE THREAD

**Issue by karlcz** [informatics-isi-edu/chaise#2073]
We would like to have a UI extension concetually related to permalink and citation. The UX is something like the following.

## User story 1: Save a query

1. Operate recordset to configure desired facet filters
2. (If configured) trigger a "save query" action
3. See a modal with simple data-entry for query metadata pre-filled with sensible defaults and a [Save] button
4. Return to recordset with some indication that current query is saved

The metadata should include
- A short query name (i.e. a saved query row name)
- A description (i.e. a longer text field)
- The author's webauthn ID (the current user performing the save, not editable)
- An extra checkbox for making it public (visible to others)?

## User story 2: Reuse a saved query

1. Operate recordset
2. (If configured) see an indication that saved queries are available on this table
3.  Trigger some kind of saved query picker
4. Pick a query
5. Return to recordet with saved filters applied nd some indication that the current query is saved

Side effect: send some event update to allow statistics to be gathered for how many times a saved query is reused.

## User story 3: Implicit matching/reuse

This may be nice to have but may be too challenging for the initial development.

1. Operate recordset
2. (If configured) app detects when facet configurations match an existing saved query
3. App automatically changes visual status to show that a saved query is in use

Side effect: send some event update to track query statistics. Set a flag to distinguish explicit vs implicit reuse?
Side effect tracked in issue #2076 .

## User story 4: Browsing of saved query storage

This may not require Chaise features other than for intra-page navigation. Visit the saved queries table as a recordset and browse the queries visible to you.  Have some way to navigate back to a result set representing that query?

## Other requirements

The configuration for this feature needs to bind the query persistence to an ermrest table. 
- The query persistence may be disabled for a deployment (default for initial prototyping)
- The query persistence may be directed to an appropriate table in the current catalog
- The query persistence may be directed to an appropriate table in a _different_ catalog
- The query save/reuse UX may be constrained to individual catalogs (default?) or allowed to share between catalogs as a config option (per persistence?  per origin?)

To support the remote catalog scenario, we probably want to include additional system metadata
- Origin server of recordset data source
- Catalog id of recordset data source
- Schema and table name of recordset data source

If sharing of queries between catalogs is enabled, can User story 4 allow the user to choose which catalog to target with a saved query?  Or should outbound navigation only take you to the originating catalog, while you have to visit other recordsets and choose query from list in order to apply to alternate data sources?

This issue depends on #2072 to allow the admin to design appropriate facet key material for persistent and cross-catalog sharing of saved queries.

**Comment by carlkesselman** [comment]
Should the query include the snapshot ID?

Also, it would be nice if we generated a MINID for the query, or some other identifier.

Carl

----------------------------------------------------------

Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute Viterbi School of Engineering Professor,

Preventive Medicine Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
 Email:  ***@***.***
Web: http://www.isi.edu/~carl
On Mar 31, 2021, 11:59 AM -0700, Karl Czajkowski ***@***.***>, wrote:

We would like to have a UI extension concetually related to permalink and citation. The UX is something like the following.

User story 1: Save a query

  1.  Operate recordset to configure desired facet filters
  2.  (If configured) trigger a "save query" action
  3.  See a modal with simple data-entry for query metadata pre-filled with sensible defaults and a [Save] button
  4.  Return to recordset with some indication that current query is saved

The metadata should include

  *   A short query name (i.e. a saved query row name)
  *   A description (i.e. a longer text field)
  *   The author's webauthn ID (the current user performing the save, not editable)
  *   An extra checkbox for making it public (visible to others)?

User story 2: Reuse a saved query

  1.  Operate recordset
  2.  (If configured) see an indication that saved queries are available on this table
  3.  Trigger some kind of saved query picker
  4.  Pick a query
  5.  Return to recordet with saved filters applied nd some indication that the current query is saved

Side effect: send some event update to allow statistics to be gathered for how many times a saved query is reused.

User story 3: Implicit matching/reuse

This may be nice to have but may be too challenging for the initial development.

  1.  Operate recordset
  2.  (If configured) app detects when facet configurations match an existing saved query
  3.  App automatically changes visual status to show that a saved query is in use

Side effect: send some event update to track query statistics. Set a flag to distinguish explicit vs implicit reuse?

User story 4: Browsing of saved query storage

This may not require Chaise features other than for intra-page navigation. Visit the saved queries table as a recordset and browse the queries visible to you. Have some way to navigate back to a result set representing that query?

Other requirements

The configuration for this feature needs to bind the query persistence to an ermrest table.

  *   The query persistence may be disabled for a deployment (default for initial prototyping)
  *   The query persistence may be directed to an appropriate table in the current catalog
  *   The query persistence may be directed to an appropriate table in a different catalog
  *   The query save/reuse UX may be constrained to individual catalogs (default?) or allowed to share between catalogs as a config option (per persistence? per origin?)

To support the remote catalog scenario, we probably want to include additional system metadata

  *   Origin server of recordset data source
  *   Catalog id of recordset data source
  *   Schema and table name of recordset data source

If sharing of queries between catalogs is enabled, can User story 4 allow the user to choose which catalog to target with a saved query? Or should outbound navigation only take you to the originating catalog, while you have to visit other recordsets and choose query from list in order to apply to alternate data sources?

**Comment by karlcz** [comment]
Maybe that should be a configurable option.  The known use case is the opposite though: it is NOT about capturing a citable set of results but about saving a recipe to apply to get new results from an evolving catalog or even a separate catalog that has a compatible model.

**Comment by ljpearlman** [comment]
Should users be able to modify and/or delete their saved queries?

**Comment by karlcz** [comment]
Delete: sure.  Perhaps just from the browsable query storage and existing row delete UX.

Modify name/description: sure.  Perhaps just from the browsable query storage and recordedit.

Modify filters: not sure.  This would seem to require a weirder UX to indicate "in saved query X with modifications" as a state in recordset, and then a revised [save] button that just updates the facet blob.  An alternate approach would be to just keep offering the normal save dialog and the user needs to purge older ones if they accumulate too many saved with same or similar names and descriptions?

**Comment by ljpearlman** [comment]
I think just letting the user delete old versions is probably sufficient (or, to get fancy, adding a "clone" operation to make it easier to create the new version before deleting the old one).

**Comment by karlcz** [comment]
You mean clone the user-entered metadata of an existing saved query while saving a new one?  Otherwise it seems like it is editing all the way down...

**Comment by karlcz** [comment]
A completely different approach (which might make sense in long term?) would be to literally offer some kind of query editor to manipulate the stored blob of facets + choices as a curation task.  This could be interesting if it eventually served other UX such as contructing more complex queries w/ logical operators that aren't' represented well by a flat facet sidebar?  But seems too ambitious to include in current goals.

**Comment by carlkesselman** [comment]
Per my previous comment, right.  We are actually talking about two separate things here: the query, which is not a resource, and the results of the query, which is.  In the past, we have used the term query to describe the second, not the first.

**Comment by carlkesselman** [comment]
Right, so a query is a resource, and a saved query is not!

Carl

----------------------------------------------------------

Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute Viterbi School of Engineering Professor,

Preventive Medicine Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
 Email:  ***@***.***
Web: http://www.isi.edu/~carl
On Mar 31, 2021, 12:12 PM -0700, Karl Czajkowski ***@***.***>, wrote:

Maybe that should be a configurable option. The known use case is the opposite though: it is NOT about capturing a citable set of results but about saving a recipe to apply to get new results from an evolving catalog or even a separate catalog that has a compatible model.

**Comment by karlcz** [comment]
Or, rather, a query is a query and a query result is a query result.

Both are resources but different kinds.  The web is very embracing
and anything can be a resource if you give it a name.

**Comment by jrchudy** [comment]
From discussion during chaise standup yesterday, user stories 1 and 2 will be our focus. User story 3 will be focused on at a later time and created in a separate issue.

This wasn't exactly emphasized above, but the intention is for these saved queries to be "per user" although, from the following that was mentioned above:

> An extra checkbox for making it public (visible to others)?

This would imply supporting the ability for other users (say user B, C, and D) to discover saved queries from user A. This also ties into user story 4 with the ability to search that saved query set (saved query table in database as a recordset page). If some queries are made "public" will those show up in that recordset table but not in the list of saved queries for the user?

**Comment by karlcz** [comment]
I imagine we can handle sharing via ermrest policy so that the basic
UI feature can just query the storage table and show whichever saved
queries are visible to the client.  Any UX need to provide
user-controllable filtering of the list would be future work.
